theseus/optimizer/nonlinear/nonlinear_optimizer.py

```diff
@@ -391,12 +400,22 @@ def retract_and_update_variables(
         converged_indices: torch.Tensor,
         step_size: float,
         force_update: bool = False,
+        grouped_retract: bool = True,
     ):
-        objective = self.objective
-
         var_idx = 0
         delta = step_size * delta
 
+        if not grouped_retract:
+            for var in self.linear_solver.linearization.ordering:
+                new_var = var.retract(delta[:, var_idx : var_idx + var.dof()])
+                if force_update:
+                    var.update(new_var.data)
+                else:
+                    var.update(new_var.data, batch_ignore_mask=converged_indices)
+                var_idx += var.dof()
+            return
```

Comment:
I don't think it's safe to `return`. Since after update variables, we also need to update `batched_cost_functions` https://github.com/facebookresearch/theseus/blob/c1906a9aaca42e1d0d3d22fdd55337fd66fb3bb8/theseus/optimizer/nonlinear/nonlinear_optimizer.py#L452.